# Creating outliers for WT02<img src="https://octodex.github.com/images/dojocat.jpg" width="100">


### This notebook is for creating false datapoints in order to test the autoencoder
#### If the autoencoder picks up the false datapoints, a higher reconstruction error should be present.


In [15]:
import sys
import numpy as np
import pandas as pd
import seaborn as sns
sns.set(color_codes=True)
import matplotlib.pyplot as plt


if not "../data_processing" in sys.path:
    print("Appended")
    sys.path.append("../data_processing")


import build_dataset


import sys, os
from scipy.stats import skew
from scipy.stats import kurtosis


ROOT_PATH = os.path.abspath("..").split("data_processing")[0]
print("ROOT", ROOT_PATH)
module_path = os.path.abspath(os.path.join(ROOT_PATH+"/utils/"))
if module_path not in sys.path:
    print("appended")
    sys.path.append(module_path)

import functions as f


c, p = f.color_palette()
sns.set(context='paper', style='whitegrid', palette=np.array(p))
print(ROOT_PATH)
plt.style.use('file://' + ROOT_PATH + "/utils/plotparams.rc")
import wt_data
import ff_transform


ROOT /Users/stianismar/Dropbox/gitProsjekter/master-thesis/src
/Users/stianismar/Dropbox/gitProsjekter/master-thesis/src


## Building the dataframe (not neccessary if it exists on disk)

In [16]:
# wt_instance = build_dataset.wt_data.load_instance("WTG02",load_minimal=False)
# df = build_dataset.create_rms_datasets_for_one_component(wt_instance, 'GbxHssRr;0,0102;m/s2', 
                                                         #power_threshold=2500,plot=False, bins=50, 
                                                         #plot_vertical_lines=False)

# build_dataset.save_dataframe_to_csv(df, 'GbxHssRr_RMS_power>2500_WTG02.csv')

## Reading the built, filtered, dataset

In [17]:
path = '/Volumes/OsvikExtra/VibrationData/RMS_dataset/GbxHssRr_RMS_power>2500_WTG02.csv'
df = pd.read_csv(path)
df.describe()

AvgPower     ActPower  AvgRotSpeed   WindSpeed  NacelleDirection  \
count   286.000000   286.000000   286.000000  286.000000        286.000000   
mean   2810.722013  2729.287770  1469.740586   11.248252        194.115734   
std     273.452548   537.939339    59.670196    2.860017         80.882321   
min    2500.165283   -37.700001   662.650623    6.600000          0.900000   
25%    2584.843689  2430.599976  1459.382731    9.400000        130.399994   
50%    2713.354126  2793.950073  1473.138835   10.400000        162.199997   
75%    2995.708069  3283.775024  1489.145768   12.500000        262.349991   
max    3315.529297  3469.699951  1562.194519   25.900000        359.299988   

       RMS_component_interval  GbxHssRr_RMS_0  GbxHssRr_RMS_1  GbxHssRr_RMS_2  \
count              286.000000      286.000000      286.000000      286.000000   
mean                 6.100239        0.457509        0.254320        0.981178   
std                  1.652317        0.045476        0.044607        0.295316   
min                  0.817078        0.249578        0.171429        0.315222   
25%                  5.267047        0.423299        0.222723        0.806082   
50%                  6.159974        0.458744        0.255291        0.951735   
75%                  7.064718        0.494233        0.277727        1.124287   
max                 14.584366        0.566679        0.568263        3.365173   

       GbxHssRr_RMS_3  ...  GbxHssRr_RMS_40  GbxHssRr_RMS_41  GbxHssRr_RMS_42  \
count      286.000000  ...       286.000000       286.000000       286.000000   
mean         0.360376  ...         1.137108         2.089491         1.875750   
std          0.066903  ...         0.139017         0.234858         0.230720   
min          0.275881  ...         0.136682         0.239168         0.238460   
25%          0.344742  ...         1.075360         1.987600         1.773638   
50%          0.355108  ...         1.125399         2.120465         1.917171   
75%          0.365520  ...         1.197883         2.221437         2.011429   
max          1.171859  ...         1.598388         2.635439         2.533416   

       GbxHssRr_RMS_43  GbxHssRr_RMS_44  GbxHssRr_RMS_45  GbxHssRr_RMS_46  \
count       286.000000       286.000000       286.000000       286.000000   
mean          2.392073         2.823761         2.271125         1.069254   
std           0.412856         0.750731         0.608550         0.263564   
min           0.219998         0.280279         0.330837         0.133726   
25%           2.272077         2.306319         1.913431         0.879402   
50%           2.494220         2.860577         2.256783         1.081592   
75%           2.651381         3.440171         2.771768         1.296891   
max           3.186057         4.257551         3.664276         1.716812   

       GbxHssRr_RMS_47  GbxHssRr_RMS_48  GbxHssRr_RMS_49  
count       286.000000       286.000000       286.000000  
mean          0.632511         0.435221         0.246160  
std           0.157723         0.102451         0.051914  
min           0.086140         0.049666         0.031207  
25%           0.516029         0.363250         0.210995  
50%           0.623535         0.425135         0.244885  
75%           0.759244         0.506057         0.282690  
max           1.139778         0.819597         0.372926  

[8 rows x 56 columns]

## Observation
We see for WT02 that the max RMS value in the gearbox range (10-2000 Hz) is 14.584366. This datapoint is interesting to look at. Maybe filter away this one?

In [18]:
rms = df['RMS_component_interval']
df = df.drop(columns= ['RMS_component_interval'])
print(f'This is the RMS values for the interval for Gearbox:{ rms }')

This is the RMS values for the interval for Gearbox:0      6.389500
1      5.061433
2      1.857597
3      6.799224
4      6.844513
         ...   
281    6.618701
282    5.376143
283    6.597880
284    8.505770
285    2.710964
Name: RMS_component_interval, Length: 286, dtype: float64


In [19]:
rms.describe()

count    286.000000
mean       6.100239
std        1.652317
min        0.817078
25%        5.267047
50%        6.159974
75%        7.064718
max       14.584366
Name: RMS_component_interval, dtype: float64

### Finding out how many bins (bin range) span across the interval for the frequency we want (10-2000 Hz)

In [20]:
# Need to find the RMS bins that span across the frequency interval we want...
SENSOR_NAME = 'GbxHssRr;0,0102;m/s2'
BINS = 50
wt_instance = wt_data.load_instance('WTG02', load_minimal=False)
for interval in wt_instance.ten_second_intervals:
  
    print("Try block")
    try:
        vibration_signal = interval.sensor_df[SENSOR_NAME]
    except:
        continue
    ts = interval.sensor_df['TimeStamp']  # Have this as the y-axis to see how the RMS/frequencies develop
    comp_type = 'gearbox'
    avg_power = interval.op_df["PwrAvg;kW"][0]
    rot_data = interval.high_speed_rot_data

    fast = ff_transform.FastFourierTransform( vibration_signal, ts, comp_type)
    fft, time, centroid, rms, rms_bins, bin_freq = fast.fft_transform_time(rot_data,
                                                                           avg_power,
                                                                           calc_rms_for_bins=True,
                                                                 plot=False,
                                                                 bins=BINS,
                                                                 plot_vertical_lines=False)
    bin_range = bin_freq[1] - bin_freq[0] # Amount of frequencies in each bin
    break


Loading WTG02...
Loaded WTG02
Try block


../data_processing/ff_transform.py:216: FutureWarning: Using a non-tuple sequence for multidimensional indexing is deprecated; use `arr[tuple(seq)]` instead of `arr[seq]`. In the future this will be interpreted as an array index, `arr[np.array(seq)]`, which will result either in an error or a different result.
  amp = fft_modulus_norm[filter_indexes]


In [21]:
bin_range

255.99902343377465

## Starting to create the "fake" dataset
- Start by taking the top 10 rows from the old dataset

In [22]:
new_dataset = df.head(10)
new_dataset

AvgPower     ActPower  AvgRotSpeed  WindSpeed  NacelleDirection  \
0  2818.207520  3291.300049  1487.871218       14.2        226.699997   
1  2528.062988  2454.500000  1471.267571       10.6        231.699997   
2  2500.473145  1772.500000  1432.996560        8.4        284.600006   
3  2555.221924  2492.899902  1465.589029        9.0        132.000000   
4  3065.522217  3304.800049  1476.364735       12.7        245.000000   
5  2579.510986  2786.899902  1469.441078       11.3        301.100006   
6  2537.796631  2466.000000  1493.188802       10.0        131.500000   
7  2588.696533  2586.899902  1493.971465       11.1        229.600006   
8  3300.607666  2901.800049  1463.211760        8.3         92.699997   
9  2534.070312  2769.000000  1498.249483        9.8        141.899994   

   GbxHssRr_RMS_0  GbxHssRr_RMS_1  GbxHssRr_RMS_2  GbxHssRr_RMS_3  \
0        0.404392        0.258909        0.811581        0.344739   
1        0.532580        0.244108        1.221325        0.346431   
2        0.478452        0.210531        1.090423        0.360744   
3        0.512141        0.226217        1.160659        0.358593   
4        0.386640        0.276154        1.027501        0.356752   
5        0.407514        0.263802        0.860637        0.369790   
6        0.517629        0.247286        0.534208        0.332621   
7        0.474903        0.279757        0.536629        0.338635   
8        0.426224        0.253064        0.872180        0.371431   
9        0.538275        0.251859        0.677084        0.340579   

   GbxHssRr_RMS_4  ...  GbxHssRr_RMS_40  GbxHssRr_RMS_41  GbxHssRr_RMS_42  \
0        1.497359  ...         1.101426         1.969857         1.778854   
1        1.477624  ...         1.106000         2.131452         1.824046   
2        0.946055  ...         0.967446         1.962500         1.463750   
3        0.996903  ...         1.068143         2.376538         1.979047   
4        1.272245  ...         1.155076         1.865711         1.766413   
5        1.132390  ...         1.160586         2.201416         1.875874   
6        1.425460  ...         1.198744         2.199523         2.190827   
7        1.641673  ...         1.338610         2.178644         1.917163   
8        0.987748  ...         1.140399         2.190679         1.957966   
9        1.732398  ...         1.235603         2.258962         2.083371   

   GbxHssRr_RMS_43  GbxHssRr_RMS_44  GbxHssRr_RMS_45  GbxHssRr_RMS_46  \
0         2.673142         3.386242         2.750154         1.419267   
1         2.082840         2.086788         1.563331         0.788437   
2         1.369159         1.247210         1.092017         0.519553   
3         2.345064         2.218439         1.995184         0.850978   
4         2.519427         3.441490         2.888474         1.341962   
5         2.447637         3.014397         2.499477         1.200119   
6         2.596931         2.703204         1.916277         1.012641   
7         2.785755         3.027529         2.128784         0.977825   
8         2.462485         2.796271         2.590859         1.137157   
9         2.619901         2.712242         1.875764         0.948645   

   GbxHssRr_RMS_47  GbxHssRr_RMS_48  GbxHssRr_RMS_49  
0         0.809299         0.444020         0.272188  
1         0.436628         0.294574         0.181487  
2         0.320830         0.221710         0.146560  
3         0.474515         0.348262         0.198423  
4         0.759380         0.538935         0.295528  
5         0.711053         0.435755         0.275101  
6         0.576606         0.344559         0.218208  
7         0.608689         0.397932         0.234199  
8         0.600736         0.447692         0.247014  
9         0.556748         0.351959         0.212861  

[10 rows x 55 columns]

### Finding out which RMS bins (features) need to be changed. num_bins_overlap

In [23]:
range_gearbox = 2000

num_bins_overlap = range_gearbox/bin_range
num_bins_overlap = int(np.ceil(num_bins_overlap))
print(f'Number of overlapping bins {num_bins_overlap}')

print(num_bins_overlap*bin_range)
print(f"Getting an additional {num_bins_overlap * bin_range - range_gearbox} frequencies...")

Number of overlapping bins 8
2047.9921874701972
Getting an additional 47.99218747019722 frequencies...


In [24]:
new_dataset = df.head(10)
new_dataset = new_dataset.drop('RMS_component_interval')
new_dataset

KeyError: "['RMS_component_interval'] not found in axis"

In [ ]:
colum_names = new_dataset.columns.values.tolist()
filtered_columns = [x for x in colum_names if '_RMS_' in x]

filtered_columns = filtered_columns[:num_bins_overlap]

In [ ]:
total_rms_array = np.zeros(num_bins_overlap)
len(total_rms_array)

In [ ]:
from IPython.display import Markdown as md

import matplotlib.pyplot as plt
# Find the sum of the 8 first RMS vals:

def calc_rms(dataframe,num_bins_overlap):
    total_rms_array = np.zeros(num_bins_overlap)
    for index, row in dataframe.iterrows():
        rms_array = []
        rms_sum = 0
        for col in filtered_columns:
            rms_sum += row[col]
            rms_array.append(row[col])
        print(f"Row: {index} The sum for the first {num_bins_overlap} rms is: {rms_sum}")
        total_rms_array = [x + y for x, y in zip(total_rms_array,rms_array)]


    avg_array = [x/len(dataframe) for x in total_rms_array] 
    fig = plt.figure(figsize=(15,5))    
    plt.plot(avg_array)
    plt.title(f"Average RMS values for the first 8 bins for {dataframe.shape[0]} rows")
    plt.margins(0)
    plt.show

    print(avg_array)

### Inspecting the RMS values for the first n overlapping bins in the original dataset

In [ ]:
calc_rms(df,int(num_bins_overlap))

## The plan 
The plan here is to update the first 8 RMS bin values based on the average seen above, and keep the remaining {{BINS-num_bins_overlap}}


In [ ]:
copy_new_df = new_dataset
for i in range(copy_new_df.shape[0]):
    for j in range(num_bins_overlap):
        copy_new_df.at[i,filtered_columns[j]] = copy_new_df.at[i,filtered_columns[j]]*2
calc_rms(copy_new_df, num_bins_overlap)
copy_new_df

In [ ]:
copy_new_df.to_csv('fake_data_points.csv',)

In [ ]:
df

# Comparing the mean RMS for the real datapoints with the fake ones

In [ ]:
import matplotlib.ticker as ticker

def plot_mean_rms_for_all_bins(df,real_or_fake,fig=None,ax=None):
    plot_columns = df.columns.values.tolist()
    plot_filtered_columns = [x for x in plot_columns if '_RMS_' in x]

    FROM = plot_filtered_columns[0]
    TO = plot_filtered_columns[-1]

    #print(plot_filtered_columns[0])
    df1 = df.loc[ :, FROM : TO ]
    plot_mean = np.mean(df1)

    if (fig == None) or (ax == None):
        fig, ax = plt.subplots()
        plt.plot(plot_mean.values.tolist(),color='r')
        
    else:
        plt.plot(plot_mean.values.tolist())
    
    
    a = np.arange(1,51,1)
    plt.margins(0)
    plt.ylim(0,np.max(plot_mean) + 1)

    for tick in ax.xaxis.get_major_ticks():
        tick.label.set_fontsize(10) 
        # specify integer or one of preset strings, e.g.
        #tick.label.set_fontsize('x-small') 
        tick.label.set_rotation('vertical')

    start, end = ax.get_xlim()
    plt.title(f"Average RMS for real and artifical datapoints")
    ax.set_ylabel("Normalized RMS value")
    ax.set_xlabel("Bin number")
    ax.xaxis.set_ticks(np.arange(1, 50, 1))
    return fig,ax
    


    #plt.plot(plot_mean.values)
    #plt.xticks(np.arange(1,50,1))

    #plt.plot(df.plot_filtered_columns)

fig2, ax2 = plot_mean_rms_for_all_bins(copy_new_df,'FAKE',fig1,ax1)
fig1, ax1 = plot_mean_rms_for_all_bins(df,'REAL')
plt.legend(['Average REAL RMS plot for 286 datapoints','Average Artificial RMS plot for 10 datapoints'],fontsize='10')
plt.show()